oci/spec_opts_windows.go

```diff
@@ -100,3 +100,32 @@ func escapeAndCombineArgs(args []string) string {
 	}
 	return strings.Join(escaped, " ")
 }
+
+// WithWindowsDevice adds a device exposed to a Windows (WCOW or LCOW) Container
+func WithWindowsDevice(IDType, ID string) SpecOpts {
+	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
+		if len(IDType) == 0 {
+			return errors.New("missing IDType")
+		}
+		if s.Windows == nil {
+			s.Windows = &specs.Windows{}
+		}
+		s.Windows.Devices = append(s.Windows.Devices, specs.WindowsDevice{IDType: IDType, ID: ID})
+		return nil
+	}
+}
+
+// WithLCOWUVMDevice adds a device exposed from the LCOW kernel to an LCOW Container
+func WithLCOWUVMDevice(path string) SpecOpts {
+	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
+		if len(path) == 0 {
+			return errors.New("missing path")
+		}
+		if s.Linux == nil {
+			return errors.New("spec is not LCOW")
+		}
```

Comment:
I'm open to feedback on this decision particularly. I think the usual idiom for these `WithX` functions is to create all the necessary structs, but I have also been bitten by uncareful `WithX` calls turning my nice, WCOW Spec silently into an LCOW Spec, and would much rather have seen a failure at spec-generate time than have lost a bunch of time trying to work out why the `windows-lcow` snapshotter was active.

Reply:
@dcantah - Would love your feedback on this one. The way I see it it's not really possible for the caller to configure an LCOW device in a meaningful way without being able to predict where that device lands. This is also true for WCOW for example. We don't configure the guest device from the perspective of the caller. You simply add the device you want and assume that the shim will figure it out. I'm not sure how you would signify the guest path and permissions for LCOW since Windows devices show up very different in the guest.

Reply:
Yea.. This is kinda tricky. Here's what we end up doing with LCOW devices in the guest to find the dev node it shows up under for example. https://github.com/microsoft/hcsshim/blob/d512c703e577ee85c4eabcffbe49f0d01f1e1af0/internal/guest/runtime/hcsv2/spec_devices.go#L133. fyi @katiewasnothere for our devices expert, I'd trust her judgement here more than mine.

Reply:
@jterry75 I agree with what you're saying here with the exception of scenarios where we want to add in dev nodes from the UVM that we know of ahead of time. For example, in the `/dev/fuse` case, there's no work from the shim side to setup or try to locate that device, it just already exists in the LCOW UVM. 

Now for the assigned devices scenario, the issue is as Justin described: we first assign the device into the guest UVM and then into the container. For LCOW, we can predict roughly what sys device path will show up for the device in the guest. However, given that drivers are what create the actual dev nodes for a device generally, there's no way for us to predict what that dev node path is going to be in the guest ahead of time.

Reply:
Here's the code we added for figuring out what dev node maps to an assigned device in the guest for reference https://github.com/microsoft/hcsshim/pull/1215. It's not pretty and I don't think it's used by anyone today.

Reply:
My expectation/intention for this code was indeed that it was only for known-location devices like `/dev/fuse` (or perhaps _only_ `/dev/fuse`), and that anything else was going to be handled by the `addAssignedDevice` code mapping a WindowsDevice entry (which exposed the device into the UVM) into a LinuxDevice entry (to expose the UVM-visible device into the container).

That said, and relating to [kevpar's comment below](https://github.com/containerd/containerd/pull/6618#pullrequestreview-905555749), I think I'll just drop this function, since it doesn't really bring anything that calling code can't do itself, and presumably with more-precise knowledge of what's happening in the container runtime, e.g. the caller would know better than this code whether the resulting spec is going to hcssim+OpenGCS, or being forwarded to a containerd instance in WSL2, etc.

The one caller I do have (the CRI implementation) does not call this code, because it really doesn't know either. ^_^